modin/numpy/arr.py

```diff
@@ -0,0 +1,1488 @@
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+"""Module houses ``array`` class, that is distributed version of ``numpy.array``."""
+
+from math import prod
+import numpy
+from pandas.core.dtypes.common import is_list_like, is_numeric_dtype, is_bool_dtype
+from pandas.api.types import is_scalar
+from inspect import signature
+
+import modin.pandas as pd
+from modin.error_message import ErrorMessage
+from modin.core.dataframe.algebra import (
+    Map,
+    Reduce,
+    Binary,
+)
+
+
+_INTEROPERABLE_TYPES = (pd.DataFrame, pd.Series)
+
+
+def try_convert_from_interoperable_type(obj):
+    if isinstance(obj, _INTEROPERABLE_TYPES):
+        new_qc = obj._query_compiler.reset_index(drop=True)
+        new_qc.columns = range(len(new_qc.columns))
+        obj = array(
+            _query_compiler=new_qc,
+            _ndim=2 if isinstance(obj, pd.DataFrame) else 1,
+        )
+    return obj
+
+
+def check_kwargs(order="C", subok=True, keepdims=None, casting="same_kind", where=True):
+    if order not in ["K", "C"]:
+        ErrorMessage.single_warning(
+            "Array order besides 'C' is not currently supported in Modin. Defaulting to 'C' order."
+        )
+    if not subok:
+        ErrorMessage.single_warning(
+            "Subclassing types is not currently supported in Modin. Defaulting to the same base dtype."
+        )
+    if keepdims:
+        ErrorMessage.single_warning(
+            "Modin does not yet support broadcasting between nested 1D arrays and 2D arrays."
+        )
+    if casting != "same_kind":
+        ErrorMessage.single_warning(
+            "Modin does not yet support the `casting` argument."
+        )
+    if not (
+        is_scalar(where) or (isinstance(where, array) and is_bool_dtype(where.dtype))
+    ):
+        if not isinstance(where, array):
+            raise NotImplementedError(
+                f"Modin only supports scalar or modin.numpy.array `where` parameter, not `where` parameter of type {type(where)}"
+            )
+        raise TypeError(
+            f"Cannot cast array data from {where.dtype} to dtype('bool') according to the rule 'safe'"
+        )
+
+
+def check_how_broadcast_to_output(arr_in: "array", arr_out: "array"):
+    if not isinstance(arr_out, array):
+        raise TypeError("return arrays must be of modin.numpy.array type.")
+    if arr_out._ndim == arr_in._ndim and arr_out.shape != arr_in.shape:
```

Comment:
this check is inaccurate, since it fails for a broadcastable case like a 4x2 w/ 4x1 (unless we're not yet supporting this)

```python
>>> A = numpy.array([[0,1,2,3], [4,5,6,7]])
>>> B = numpy.array([[8,9,10,11]])
>>> A + B
array([[ 8, 10, 12, 14],
       [12, 14, 16, 18]])
```

This would fall into the [broadcasting rules](https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules) case where the rightmost dimension of A and B, and the next dimension of B is 1 so that of A doesn't matter.

Reply:
Actually never mind, I think I misunderstood the purpose of this function (I thought it was comparing the dimensions of two inputs of binary operators). Based on how this function is called, is it correct to assume that `arr_in` is always the result of some computation, which may have already broadcast its inputs as necessary?

Reply:
Yup!